Paintroid/src/main/java/org/catrobat/paintroid/ui/dragndrop/DragAndDropListView.kt

```diff
@@ -0,0 +1,195 @@
+        context,
+        attributes,
+        defStyle
+    )
+
+    init {
+        onItemLongClickListener = OnItemLongClickListener { parent, view, position, id ->
+            presenter?.onLongClickLayerAtPosition(position, view)
+            true
+        }
+        onItemClickListener = OnItemClickListener { parent, view, position, id ->
+            presenter?.onClickLayerAtPosition(
+                position,
+                view
+            )
+        }
+    }
+
+    fun setPresenter(presenter: DragAndDropPresenter) {
+        this.presenter = presenter
+    }
+
+    override fun onTouchEvent(event: MotionEvent): Boolean {
+        if (hoveringListItem == null) {
+            return super.onTouchEvent(event)
+        }
+        when (event.action) {
+            MotionEvent.ACTION_UP -> handleTouchUp()
+            MotionEvent.ACTION_CANCEL -> stopDragging()
+            MotionEvent.ACTION_DOWN -> downY = event.y
+            MotionEvent.ACTION_MOVE -> {
+                val dY = event.y - downY
+                downY += dY
+                downY -= offsetToCenter.toFloat()
+                viewBounds?.let {
+                    it.offset(it.left, downY.toInt())
+                    val top = min(height - it.height(), max(0, it.top))
+                    val bottom = top + it.height()
+                    hoveringListItem?.setBounds(it.left, top, it.right, bottom)
+                }
+                invalidate()
+                swapListItems()
+            }
+        }
+        return true
+    }
+
+    override fun handleOnItemLongClick(position: Int, view: View) {
+        this.view?.visibility = VISIBLE
+        this.view = view
+        initialPosition = position
+        this.position = position
+        view.visibility = INVISIBLE
+        hoveringListItem = getHoveringListItem(view)
+        setOffsetToCenter(viewBounds)
+        invalidate()
+    }
+
+    override fun stopDragging() {
+        if (hoveringListItem != null) {
+            mergePosition = -1
+            view?.visibility = VISIBLE
+            view = null
+            hoveringListItem = null
+            invalidate()
+        }
+    }
+
+    public override fun dispatchDraw(canvas: Canvas) {
+        super.dispatchDraw(canvas)
+        hoveringListItem?.draw(canvas)
+    }
+
+    private fun getHoveringListItem(view: View): BitmapDrawable {
+        val bitmap = Bitmap.createBitmap(view.width, view.height, Bitmap.Config.ARGB_8888)
+        val canvas = Canvas(bitmap)
+        view.draw(canvas)
+        val drawable = BitmapDrawable(resources, bitmap)
+        viewBounds = Rect(view.left, view.top, view.right, view.bottom)
+        viewBounds?.let {
+            drawable.bounds = it
+        }
+        drawable.alpha = 192
+        return drawable
+    }
+
+    private fun setOffsetToCenter(viewBounds: Rect?) {
+        if (viewBounds != null) {
+            offsetToCenter = viewBounds.height() / 2
+        }
+    }
+
+    private fun swapListItems() {
```

Comment:
Please try to make this method clean :)

Reply:
done, I've tried to clean and simplify it as much as I could, please take look at it again :)